core/Derive.carp

```diff
@@ -0,0 +1,92 @@
+(doc Derive "is a mechanism for deriving interfaces automatically.
+
+Please reference [the documentation](https://github.com/carp-lang/Carp/blob/master/docs/Derive.md)
+for more information.")
+(defmodule Derive
+  (hidden derivers)
+  (defdynamic derivers '())
+
+  (doc make-deriver "is a mechanism for providing your own deriver by providing
+the interface name `f`, the arguments it takes `args`, and a function that can
+generate a body when given a type `body`.
+
+Example:
+```
+(make-deriver 'zero []
+  (fn [t]
+    (cons 'init
+      (map (fn [_] '(zero)) (members t)))))
+```")
+  (defndynamic make-deriver [f args body]
+    (set! Derive.derivers
+      (cons
+        (list f
+          (fn [t]
+            (list 'defmodule t
+              (list 'defn f args
+                (body t))
+              (list 'implements f (Symbol.prefix t f)))))
+        Derive.derivers)))
+
+  (doc make-update-deriver "is a convenience function for creating a deriver
+for functions that you could pass into `update-<member>` style functions.")
+  (defndynamic make-update-deriver [interface]
+    (make-deriver interface ['o]
+      (fn [t]
+        (reduce
+          (fn [acc m]
+            (list (Symbol.concat ['update- (car m)]) acc (list 'ref interface)))
+          'o
+          (eval (list 'members t))))))
+
+  (hidden get-deriver)
+  (defndynamic get-deriver [f derivers]
+    (if (empty? derivers)
+      '()
+      (if (= (caar derivers) f)
+        (car derivers)
+        (get-deriver f (cdr derivers)))))
+
+  (doc derivable? "checks whether a quoted interface name `f` is currently
+derivable.")
+  (defndynamic derivable? [f]
+    (let [deriver (get-deriver f Derive.derivers)]
+      (not (empty? deriver))))
+
+  (doc derivables "returns the list of currently derivable interfaces.")
+  (defndynamic derivables []
+    (map car Derive.derivers))
+
+  (doc derive "derives an interface function `f` for a type `t`.")
+  (defmacro derive [t f]
```

Comment:
I don't think it's totally necessary, but I feel this might read a bit better if `t` is the second argument:

`(derive zero Point)`

This also would make it follow the order of `implements`

`(implements zero foo)`

Reply:
Fair enough; what’s the consensus here, @carp-lang/maintainers?

Reply:
I think it makes a lot of sense to have the same order for derive & implements. Since we already have a lot of code that relies on the `implements` ordering it seems reasonable to go with that.

Reply:
Always felt that the order in `implements` should be the other way around, especially if you read it "infix": `zero implements foo`. But there is certainly something to be said for consistency.

Reply:
> Always felt that the order in implements should be the other way around, especially if you read it "infix": zero implements foo.
But there is certainly something to be said for consistency.

That's a fair point, if others feel this way too, I'm ok with making a change on the `implements` side. The reason it has the order is because when considering prefix notation, `(implements zero foo)` made more sense to me--basically imagine there's a comma or colon when reading it off as english-- `implements zero: foo`--it also allows the form to take varargs in the future `(implements zero foo...)`. 

`(implements foo zero)`, putting the implementation first, felt odd and seemed more likely to create friction when skimming code even if it corresponds to the correct infix order. At the very least, I'd personally assume `foo` was the interface in that case, why precisely I don't know but it's how my brain wants to naturally parse it (this may well be a personal problem, so I'm happy to go along with consensus).

Another argument for changing the order on the `implements` side would be that it's more consistent, I think, with some of our other forms, `meta` for example.

Reply:
It's definitely not super clear which one is most natural. I usually don't try to make it "read naturally" in the natural language kind of way. Instead (even when now programming Haskell) I use some kind of "currying" mindset where the bigger or more stable argument comes first to the function. This is of course kind of vague, but most obvious example is maybe a setter; the thing being modified comes first and then comes the new value.

Applying the same idea to the functions we're discussing seem to suggest that `implements` should switch, since we're "modifying that type". One could maybe also say that we're modifying the interface, but I don't think that's how people generally think about it (even though it's kind of what's happening in the compiler internals.)

All this is to say that I kind of agree that this PR has the superior order, but it's a close call and I didn't want to suggest changing all the other code in the same PR. 

Maybe we can just merge this as-is and fix the ordering later? That will give us more time to think about it, without blocking people from trying out `derive`.

Reply:
I see merits in either ordering, and I don’t mind either. What would be interesting would be to enable people to derive multiple interfaces in one call:

```clojure
(derive [= zero str] T)
; vs
(derive T [= zero str])
```

Where, for me, the second ordering seems more natural; this of course is entirely subjective as well.

I’d argue that for now keeping concistency and not breaking existent behavior probably trumps any need to change, meaning that I should go ahead and switch the order for `derive` and we can revisit this another time. How does that sound?

Reply:
Deriving multiple ones sounds great. Could be a `*` macro also, of course (or just rest args?)

> How does that sound?

Perfect!

Reply:
While changing the order I realized that this would also change overrides to look like this:

```clojure
(derive = T eq) ; => derive = for T as eq
```

I don’t know whether I like that very much.

Reply:
ugh, no that's really confusing

Reply:
`(derive T [= zero str])` Cool 🙂

Reply:
Cool, good discussion! Let's keep this order and change the order of `implements` instead. I can take that on.

Reply:
Okay, I changed the order back for now then :) I’ll resolve this discussion for now!